plugin/trino-hive/src/main/java/io/trino/plugin/hive/fs/TransactionScopeCachingDirectoryLister.java

```
@@ -0,0 +1,230 @@
+    {
+        // Some Hive table types (e.g.: views) do not have a storage location
+        return storage.getOptionalLocation().isPresent() && isNotEmpty(storage.getLocation());
+    }
+
+    private static class FetchingValueHolder
+    {
+        private final List<LocatedFileStatus> cachedFiles = synchronizedList(new ArrayList<>());
+        @GuardedBy("this")
+        @Nullable
+        private RemoteIterator<LocatedFileStatus> fileIterator;
+        @GuardedBy("this")
+        @Nullable
+        private IOException exception;
+
+        public FetchingValueHolder(RemoteIterator<LocatedFileStatus> fileIterator)
+        {
+            this.fileIterator = requireNonNull(fileIterator, "fileIterator is null");
+        }
+
+        public synchronized boolean isFullyCached()
+        {
+            return fileIterator == null && exception == null;
+        }
+
+        public int getCachedFilesSize()
+        {
+            return cachedFiles.size();
+        }
+
+        public RemoteIterator<LocatedFileStatus> getCachedFiles()
+        {
+            checkState(isFullyCached());
+            return new SimpleRemoteIterator(cachedFiles.iterator());
+        }
+
+        public Optional<LocatedFileStatus> getCachedFile(int index)
+                throws IOException
+        {
+            int filesSize = cachedFiles.size();
+            checkArgument(index >= 0 && index <= filesSize);
+
+            if (index < filesSize) {
+                return Optional.of(cachedFiles.get(index));
+            }
+
+            return fetchNextCachedFile(index);
+        }
+
+        private synchronized Optional<LocatedFileStatus> fetchNextCachedFile(int index)
+                throws IOException
+        {
+            if (index < cachedFiles.size()) {
+                // file was fetched concurrently
+                return Optional.of(cachedFiles.get(index));
+            }
+
+            if (exception != null) {
```

Check this first, before trying to pull an entry from the cache -- fail fast